kingpin/actors/aws/s3.py

```diff
@@ -334,461 +457,363 @@ def _generate_lifecycle(self, config):
-            mock_bucket.get_lifecycle_config.side_effect = S3ResponseError(
-                404, 'Empty')
-
-            raise gen.Return(mock_bucket)
-
-        # This throws no exceptions, even if the bucket exists, that we know
-        # about or can expect.
-
-        # https://github.com/boto/boto3/issues/125
-        location_args = {}
+        params = {
+            'Bucket': self.option('name')
+        }
+
         if self.option('region') != 'us-east-1':
-            location_args = {'location': self.option('region')}
+            params['CreateBucketConfiguration'] = {
+                'LocationConstraint': self.option('region')
+            }
 
         self.log.info('Creating bucket')
-        bucket = yield self.thread(self.s3_conn.create_bucket,
-                                   self.option('name'), **location_args)
-        raise gen.Return(bucket)
+        yield self.thread(self.s3_conn.create_bucket, **params)
 
     @gen.coroutine
-    def _verify_can_delete_bucket(self, bucket):
+    def _verify_can_delete_bucket(self):
         # Find out if there are any files in the bucket before we go to delete
         # it. We cannot delete a bucket with files in it -- nor do we want to.
-        keys = yield self.thread(bucket.get_all_keys)
-        if len(keys) > 0:
+        bucket = self.option('name')
+        keys = yield self.thread(self.s3_conn.list_objects, Bucket=bucket)
+
+        if 'Contents' not in keys:
+            raise gen.Return()
+
+        if len(keys['Contents']) > 0:
             raise exceptions.RecoverableActorFailure(
                 'Cannot delete bucket with keys: %s files found' % len(keys))
 
     @gen.coroutine
-    @dry('Would have deleted bucket {bucket}')
-    def _delete_bucket(self, bucket):
-        """Tries to delete an S3 bucket.
-
-        args:
-            bucket: The S3 bucket object as returned by Boto
-        """
+    @dry('Would have deleted bucket')
+    def _delete_bucket(self):
+        bucket = self.option('name')
         try:
             self.log.info('Deleting bucket %s' % bucket)
-            yield self.thread(bucket.delete)
-        except S3ResponseError as e:
-            if e.status == 409:
-                raise exceptions.RecoverableActorFailure(
-                    'Cannot delete bucket: %s' % e.message)
+            yield self.thread(self.s3_conn.delete_bucket, Bucket=bucket)
+        except ClientError as e:
+            raise exceptions.RecoverableActorFailure(
+                'Cannot delete bucket: %s' % e.message)
 
     @gen.coroutine
-    def _ensure_policy(self, bucket):
-        """Ensure the policy attached to the bucket is correct.
+    def _get_policy(self):
+        try:
+            raw = yield self.thread(
+                self.s3_conn.get_bucket_policy,
+                Bucket=self.option('name'))
+            exist = json.loads(raw['Policy'])
+        except ClientError as e:
+            if 'NoSuchBucketPolicy' in e.message:
```

Comment:
You're missing the `else: raise` here. Very important! Otheriwse all ClientError that is not NoScuh will be simply ignored